x/auth/ante.go

```diff
@@ -9,82 +9,95 @@ func NewAnteHandler(accountMapper sdk.AccountMapper) sdk.AnteHandler {
-		var msg = tx.GetMsg()
-		var signerAddrs = msg.GetSigners()
-		var signerAccs = make([]sdk.Account, len(signerAddrs))
+		// TODO: can tx just implement message?
+		msg := tx.GetMsg()
 
 		// Assert that number of signatures is correct.
+		var signerAddrs = msg.GetSigners()
 		if len(sigs) != len(signerAddrs) {
 			return ctx,
 				sdk.ErrUnauthorized("wrong number of signers").Result(),
 				true
 		}
 
-		// Check each nonce and sig.
-		// TODO Refactor out.
-		for i, sig := range sigs {
-
-			var signerAcc = accountMapper.GetAccount(ctx, signerAddrs[i])
-			signerAccs[i] = signerAcc
+		// Collect accounts to set in the context
+		var signerAccs = make([]sdk.Account, len(signerAddrs))
 
-			// If no pubkey, set pubkey.
-			if signerAcc.GetPubKey().Empty() {
-				err := signerAcc.SetPubKey(sig.PubKey)
-				if err != nil {
-					return ctx,
-						sdk.ErrInternal("setting PubKey on signer").Result(),
-						true
-				}
-			}
+		// First sig is the fee payer.
+		// signBytes uses the sequence of the fee payer
+		// (ie. the first account)
+		payerAddr, payerSig := signerAddrs[0], sigs[0]
+		signBytes := sdk.StdSignBytes(ctx.ChainID(), payerSig.Sequence, msg)
 
-			// Check and increment sequence number.
-			seq := signerAcc.GetSequence()
-			if seq != sig.Sequence {
-				return ctx,
-					sdk.ErrInvalidSequence("").Result(),
-					true
-			}
-			signerAcc.SetSequence(seq + 1)
+		// Check fee payer sig and nonce, and deduct fee.
+		// This is done first because it only
+		// requires fetching 1 account.
+		payerAcc, res := processSig(ctx, accountMapper, payerAddr, payerSig, signBytes)
+		if !res.IsOK() {
+			return ctx, res, true
+		}
+		signerAccs[0] = payerAcc
+		// TODO: Charge fee from payerAcc.
+		// TODO: accountMapper.SetAccount(ctx, payerAddr)
 
-			// Check sig.
-			if !sig.PubKey.VerifyBytes(msg.GetSignBytes(), sig.Signature) {
-				return ctx,
-					sdk.ErrUnauthorized("").Result(),
-					true
+		// Check sig and nonce for the rest.
+		for i := 1; i < len(sigs); i++ {
+			signerAddr, sig := signerAddrs[i], sigs[i]
+			signerAcc, res := processSig(ctx, accountMapper, signerAddr, sig, signBytes)
+			if !res.IsOK() {
+				return ctx, res, true
 			}
-
-			// Save the account.
-			accountMapper.SetAccount(ctx, signerAcc)
+			signerAccs[i] = signerAcc
 		}
 
 		ctx = WithSigners(ctx, signerAccs)
 		return ctx, sdk.Result{}, false // continue...
 	}
 }
+
+// verify the signature and increment the sequence.
+// if the account doesn't have a pubkey, set it as well.
+func processSig(ctx sdk.Context, am sdk.AccountMapper, addr sdk.Address, sig sdk.StdSignature, signBytes []byte) (acc sdk.Account, res sdk.Result) {
+
+	// Get the account
+	acc = am.GetAccount(ctx, addr)
+	if acc == nil {
+		return nil, sdk.ErrUnrecognizedAddress(addr).Result()
+	}
+
+	// Check and increment sequence number.
+	seq := acc.GetSequence()
+	if seq != sig.Sequence {
+		return nil, sdk.ErrInvalidSequence("").Result()
+	}
+	acc.SetSequence(seq + 1)
```

Comment:
Incrementing sequence numbers of all signers accounts instead of just incrementing the fee payers sequence number decreases usability. 

It can cause an unrelated transaction by one of the co-signers to fail unexpectedly if the multisig tx is delivered.

Reply:
So we should only require the nonce for the fee payer and only increment the nonce for the fee payer ?

Reply:
yes. Unless I am missing something, there are no replay vulnerabilities if one nonce increments because the txsig from each sender should coverall the txs in a bundle